lxd/db/profiles.go

```diff
@@ -47,15 +47,15 @@ type Profile struct {
 	Description string `db:"coalesce=''"`
 	Config      map[string]string
 	Devices     map[string]Device
-	UsedBy      []string
+	Instances   []string `db:"omit=create,update&uri=yes"`
```

Comment:
@masnax im interested what was your thinking around this change? To the naive reader it seemed like having it called `UsedBy` aligned better.

Reply:
`profiles.UsedBy` only holds instance URIs at the moment, while `projects.UsedBy` contains instances, networks, acls, storage volumes, etc. This leaves two options with how to deal with this on the generator side:
* Use `UsedBy []string` with tags like `db:"used-by=instances"` to specify what the slice contains
* Just split apart `UsedBy` into its component entity slices: `Instances', `Networks`, etc.

I went with the latter, although as Networks, NetworkACLs, and Storage Volumes are as yet unsupported by the generator, I left UsedBy in place for projects. I also think it's more appropriate as the database shouldn't have any concept of what `UsedBy` is, and instead we can just fill a `UsedBy` slice when converting to an API struct.

Reply:
I see, I think. TBH I'm not really totally happy with us having multiple ways of generating the `UsedBy` values (as we to today already), some places it uses in-Go logic and multiple queries and others its done at object load time apparently (although I'm struggling to see where the queries are actually done due to this being DB generation logic).

Recently I've been moving away from populating UsedBy at object load time from the DB and instead doing it at render time, as often we don't need to go to the expense of populating UsedBy unless its actually being asked for.

Where can I see the resulting query that this logic implements? Thanks

Reply:
So there's 3 parts to this:
* Presence of `profile.Instances` 
  * this makes the generated methods for `Get/Create/UpdateProfile` look for their associated `Get/Create/UpdateProfileInstances` methods. These methods would be defined in `instances_profiles.mapper.go` and be generated by the line `//go:generate mapper method -p db -e instance_profile GetMany struct=Profile` as an example for `GetProfileInstances`.
* `db:omit` tags
  * This prevents `CreateProfile` and `UpdateProfile` from detecting `profile.Instances`, because we don't want the profile to change what its used by.
* `db:uri` tag
  * This tells any of the generated methods that use `profile.Instances` to keep the URI format, and not parse it for just the name.

But to directly answer your question:
* `GetProfiles` in `profiles.mapper.go` calls `GetProfileInstances` in `instance_profiles.mapper.go`, and then calls `GetInstanceURIs` using the `instance_id`s obtained.

Reply:
>     * `GetProfiles` in `profiles.mapper.go` calls `GetProfileInstances` in `instance_profiles.mapper.go`, and then calls `GetInstanceURIs` from `instances.mapper.go` using the `instance_id`s obtained

Great thanks that clears things up for me. So this doesn't sound particularly more efficient that what, say, what we do for ACL's UsedBy:

https://github.com/lxc/lxd/blob/9b84e70a0f0e2bb513f480b7138aaa690dbf97a4/lxd/network_acls.go#L153-L159

I wonder if we should remove the population of UsedBy from these functions and instead have a separately implemented `UsedBy` function for each object type that we can then call to populate the `UsedBy` field when needed. This would be more efficient that loading this always as often its unnecessary.